File: leetcode-in-python/solution/75_sort_colors.py
import unittest
from typing import List


class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        l, i, r = 0, 0, len(nums) - 1
        while i <= r:
            if nums[i] == 0:
                nums[l], nums[i] = nums[i], nums[l]
                l += 1
                i += 1
            elif nums[i] == 1:
                i += 1
            else:  # nums[i] == 2
                nums[i], nums[r] = nums[r], nums[i]
                r -= 1


class TestSolution(unittest.TestCase):

    def setUp(self):
        self.test_case = [
            ([2, 0, 2, 1, 1, 0], [0, 0, 1, 1, 2, 2]),
            ([2, 0, 1], [0, 1, 2]),
            ([0, 1, 0], [0, 0, 1]),
            ([0], [0]),
            ([1], [1]),
            ([2], [2]),
        ]
        self.s = Solution()

    def test_solution(self):
        for nums, answer in self.test_case:
            self.s.sortColors(nums)
            self.assertEqual(nums, answer)


if __name__ == '__main__':
    unittest.main()

File: leetcode-in-rust/src/solution/_75_sort_colors.rs
pub struct Solution {}

impl Solution {
    pub fn sort_colors(nums: &mut Vec<i32>) {
        let (mut l, mut i, mut r) = (0, 0, nums.len() - 1);
        while i <= r {
            match nums[i] {
                0 => {
                    nums.swap(l, i);
                    l += 1;
                    i += 1;
                }
                1 => i += 1,
                _ => {
                    nums.swap(r, i);
                    if r == 0 {
                        break;
                    }
                    r -= 1;
                }
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        let test_case: Vec<(Vec<i32>, Vec<i32>)> = vec![
            (vec![2, 0, 2, 1, 1, 0], vec![0, 0, 1, 1, 2, 2]),
            (vec![2, 0, 1], vec![0, 1, 2]),
            (vec![0, 1, 0], vec![0, 0, 1]),
            (vec![0], vec![0]),
            (vec![1], vec![1]),
            (vec![2], vec![2]),
        ];
        for (mut nums, answer) in test_case {
            Solution::sort_colors(nums.as_mut());
            assert_eq!(nums, answer);
        }
    }
}
